pkg/server/node.go

```diff
@@ -1044,3 +1045,81 @@ func (n *Node) RangeFeed(
 	}
 	return nil
 }
+
+func (n *Node) UnsafeHealRange(
+	ctx context.Context, req *roachpb.UnsafeHealRangeRequest,
+) (_ *roachpb.UnsafeHealRangeResponse, rErr error) {
+	defer func() {
+		if r := recover(); r != nil {
+			rErr = errors.Errorf("%v", r)
+		}
+	}()
+
+	var expValue *roachpb.Value
+	var desc roachpb.RangeDescriptor
+	if err := n.storeCfg.DB.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
+		kvs, err := sql.ScanMetaKVs(ctx, txn, roachpb.Span{
+			// TODO(tbg): paginate.
+			Key:    roachpb.KeyMin,
+			EndKey: roachpb.KeyMax,
+		})
+		if err != nil {
+			return err
+		}
+
+		for i := range kvs {
+			if err := kvs[i].Value.GetProto(&desc); err != nil {
+				return err
+			}
+			if desc.RangeID == req.RangeID {
+				expValue = kvs[i].Value
+				return nil
+			}
+		}
+		return errors.Errorf("r%d not found", req.RangeID)
+	}); err != nil {
+		return nil, err
+	}
+
+	var s *storage.Store
+	if err := n.stores.VisitStores(func(inner *storage.Store) error {
+		if s == nil {
+			s = inner
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	storeID, nodeID := s.Ident.StoreID, s.Ident.NodeID
+
+	dead := append([]roachpb.ReplicaDescriptor(nil), desc.Replicas().All()...)
+	to := desc.AddReplica(nodeID, storeID, roachpb.VOTER_FULL)
+	for _, rd := range dead {
+		desc.RemoveReplica(rd.NodeID, rd.StoreID)
+	}
+
+	if err := n.storeCfg.DB.CPut(
```

Comment:
Here we're updating the meta2 range with this new dummy descriptor. Is it safe to do this before creating the replica? I think it's fine, we're no more "unsafe" than before.

The basic primitive should also not really touch meta2. We should again rely on higher level abstractions to do this for us.

Note to self: There are two copies of the range descriptor: one in meta2, and one on the range itself. They're usually transactionally consistent. There are some invariants here around how intents are stored, how we recover from finding intents on stores, and whether to assume they're committed or not. I'm a bit fuzzy here.

Reply:
It's also fine to not update the meta2. The real source of truth is always the range local one. As written we're CPut-ing on the meta2 key. We _should_ be able to CPut on the range local one and then eventually blindly write to the meta2 one. We already rely on something similar for our offline recovery tool (where we CPut to the range local one only).

The danger is that if we don't update meta2, there's a race with the replicaGCqueue. This dummy replica may very well GC itself (after consulting the meta2 range). 

Squinting here a bit further: could a leaseholder read on the local copy be used instead of meta2? It has the nice property of avoiding putting meta2 on the hotpath for online recovery of multiple ranges.